pkg/deploy/internal/segment/segment.go

```diff
@@ -0,0 +1,192 @@
+			return "", nil
+		}
+		return "", fmt.Errorf("failed to fetch segment object: %w", err)
+	}
+	//Owner field is required in a PUT
+	value, ok := request["owner"]
+	if !ok || value == "" {
+		responseData, err := getJsonResponseFromSegmentsResponse(jsonResponse)
+		if err != nil {
+			return "", err
+		}
+		request["owner"] = responseData.Owner
+	}
+
+	request["uid"] = c.OriginObjectId
+	payload, err := json.Marshal(request)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal segment request: %w", err)
+	}
+
+	_, err = deploy(ctx, client, c.OriginObjectId, payload, c)
+	if err != nil {
+		return "", fmt.Errorf("failed API request: %w", err)
+	}
+
+	return c.OriginObjectId, nil
+}
+
+func deployWithExternalId(ctx context.Context, client DeploySegmentClient, request map[string]any, c *config.Config, externalId string) (string, error) {
+	segmentsResponses, err := client.GetAll(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to GET segments: %w", err)
+	}
+
+	var responseData jsonResponse
+	for _, segmentResponse := range segmentsResponses {
+		responseData, err = getJsonResponseFromSegmentsResponse(segmentResponse)
+		if err != nil {
+			return "", err
+		}
+		//In case of a match, the put needs additional fields
+		if responseData.ExternalId == externalId {
+			request["uid"] = responseData.UID
+			value, ok := request["owner"]
+			if !ok || value == "" {
+				request["owner"] = responseData.Owner
+			}
+			break
+		}
+	}
+
+	payload, err := json.Marshal(request)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal segment request: %w", err)
+	}
+
+	responseUpsert, err := deploy(ctx, client, responseData.UID, payload, c)
+	if err != nil {
+		return "", fmt.Errorf("failed API request: %w", err)
+	}
+
+	//For a POST we need to parse the response again to read out the ID
+	if responseUpsert.StatusCode == http.StatusCreated {
+		responseData, err = getJsonResponseFromSegmentsResponse(responseUpsert)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return responseData.UID, nil
+}
+
+func deploy(ctx context.Context, client DeploySegmentClient, id string, payload []byte, c *config.Config) (segment.Response, error) {
+	//create new context to carry logger
+	ctx = logr.NewContext(ctx, log.WithCtxFields(ctx).GetLogr())
+	ctx, cancel := context.WithTimeout(ctx, time.Minute)
+	defer cancel()
+
+	responseUpsert, err := client.Upsert(ctx, id, payload)
+	if err != nil {
+		var apiErr api.APIError
+		if errors.As(err, &apiErr) {
+			return api.Response{}, fmt.Errorf("failed to upsert segment with id %q: %w", id, err)
+		}
+
+		return api.Response{}, deployErrors.NewConfigDeployErr(c, fmt.Sprintf("failed to upsert segment with id %q", id)).WithError(err)
+	}
+
+	return responseUpsert, nil
+}
+
+func createResolveEntity(id string, externalId string, properties parameter.Properties, c *config.Config) entities.ResolvedEntity {
+	properties[config.IdParameter] = id
+	return entities.ResolvedEntity{
+		EntityName: externalId,
```

Comment:
Note to self: I think we can get rid of this `EntityName` in the future. It has only one read access, and that's in a test. Nothing to do here for now